### PR TITLE
Refactoring TFTP client

### PR DESCRIPTION
**Pull request by JeonginKim**
Added TFTP client sample application

### CODE REVIEW

File: apps/netutils/tftpc/tftpc_get.c [hunk 1]
```diff
@@ -124,9 +124,9 @@ static inline ssize_t tftp_write(int fd, const uint8_t *buf, size_t len)
 
 static inline int tftp_parsedatapacket(const uint8_t *packet, uint16_t *opcode, uint16_t *blockno)
 {
-	*opcode = (uint16_t)packet[0] << 8 | (uint16_t)packet[1];
+	*opcode = (uint16_t) packet[0] << 8 | (uint16_t) packet[1];
```
Comment (sunghan-chang) [inline apps/netutils/tftpc/tftpc_get.c]: Let's revert this line to original. This causes violations of coding rule.

File: apps/netutils/tftpc/tftpc_get.c [hunk 2]
```diff
@@ -124,9 +124,9 @@ static inline ssize_t tftp_write(int fd, const uint8_t *buf, size_t len)
...
 	if (*opcode == TFTP_DATA) {
-		*blockno = (uint16_t)packet[2] << 8 | (uint16_t)packet[3];
+		*blockno = (uint16_t) packet[2] << 8 | (uint16_t) packet[3];
```
Comment (sunghan-chang) [inline apps/netutils/tftpc/tftpc_get.c]: same as above

File: apps/system/utils/Makefile [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ###########################################################################
 #
-# Copyright 2016 Samsung Electronics All Rights Reserved.
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/system/utils/Makefile]: Why do you change this to 2017?

File: apps/system/utils/netcmd.c [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /****************************************************************************
  *
- * Copyright 2016 Samsung Electronics All Rights Reserved.
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd.c]: same as previous

File: apps/system/utils/netcmd.c [hunk 5]
```diff
@@ -79,21 +81,14 @@
...
-
-#if defined(CONFIG_NETUTILS_TFTPC)
-struct tftpc_args_s {
-	bool binary;				/* true:binary ("octet") false:text ("netascii") */
-	bool allocated;				/* true: destpath is allocated */
-	char *destpath;				/* Path at destination */
-	const char *srcpath;		/* Path at src */
-	in_addr_t ipaddr;			/* Host IP address */
-};
+#ifdef CONFIG_LWM2M_CLIENT_MODE
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd.c]: Is this a change for tftpc?

File: apps/system/utils/netcmd.c [hunk 6]
```diff
@@ -556,11 +374,9 @@ int cmd_ifconfig(int argc, char **argv)
 	return OK;
 }
 
-
-
 const static tash_cmdlist_t net_utilcmds[] = {
-#if defined(CONFIG_NETUTILS_TFTPC) && !defined(CONFIG_DISABLE_ENVIRON)
-	{"ftp_get", cmd_get, TASH_EXECMD_SYNC},
+#ifdef CONFIG_NETUTILS_TFTPC
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd.c]: Put it by Arphabet order.

File: apps/system/utils/netcmd_tftpc.c [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+#include <arpa/inet.h>
+#include "netcmd.h"
+#include <apps/netutils/tftp.h>
+
+#if defined(CONFIG_NETUTILS_TFTPC)
+struct tftpc_args_s {
+	bool binary;				/* true:binary ("octet") false:text ("netascii") */
+	bool allocated;				/* true: destpath is allocated */
+	char *destpath;				/* Path at destination */
+	const char *srcpath;		/* Path at src */
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd_tftpc.c]: How about align for comments?

File: apps/system/utils/netcmd_tftpc.c [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+	in_addr_t ipaddr;			/* Host IP address */
+};
+
+int tftpc_parseargs(int argc, char **argv, struct tftpc_args_s *args)
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd_tftpc.c]: No one calls this function in this file and this is not a public API. Is this needed?

File: apps/system/utils/netcmd.c [hunk 9]
```diff
@@ -579,6 +395,12 @@ const static tash_cmdlist_t net_appcmds[] = {
 #ifdef CONFIG_SYSTEM_NETDB
 	{"netdb", netdb_main, TASH_EXECMD_SYNC},
 #endif
+#ifdef CONFIG_LWM2M_CLIENT_MODE
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd.c]: same as previous

File: apps/system/utils/netcmd_tftpc.c [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	in_addr_t ipaddr;
+	ipaddr = inet_addr(argv[3]);
+
+	bool isBinary = false;
+	if (*argv[4] == 'b' || *argv[4] == 'B') {
+		isBinary = true;
+	}
+
+	if (tftpget(argv[1], argv[2], ipaddr, isBinary) != OK) {
+		printf(fmtcmdfailed, argv[0], "tftpget");
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd_tftpc.c]: The fmtcmdfailed has three args, but you give two args.
static const char fmtcmdfailed[] = "%s: %s failed: %s\n";

File: apps/system/utils/netcmd_tftpc.c [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	in_addr_t ipaddr;			/* Host IP address */
+};
+
+// example : tftpc tftptest.txt /mnt/tftptest.txt 192.168.1.216
+// example : tftpc tftptest.bin /mnt/tftptest.bin 192.168.1.216 b
+
+int cmd_tftpc(int argc, char **argv)
+{
+	in_addr_t ipaddr;
+	ipaddr = inet_addr(argv[3]);
```
Comment (sunghan-chang) [inline apps/system/utils/netcmd_tftpc.c]: Before using , don't you need to check validation of args? from argv[1] to argv[4]?

File: apps/examples/smtp_client/smtp_client.c [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+ * Included Files
+ ****************************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <apps/shell/tash.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <tinyara/config.h>
```
Comment (sunghan-chang) [inline apps/examples/smtp_client/smtp_client.c]: This header is not needed.

File: apps/examples/smtp_client/smtp_client.c [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <apps/shell/tash.h>
```
Comment (sunghan-chang) [inline apps/examples/smtp_client/smtp_client.c]: same as previous

File: apps/examples/smtp_client/smtp_client.c [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
```
Comment (sunghan-chang) [inline apps/examples/smtp_client/smtp_client.c]: same

File: apps/examples/smtp_client/smtp_client.c [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+#include <string.h>
+#include <apps/shell/tash.h>
+#include <net/if.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <tinyara/config.h>
+#include <apps/netutils/netlib.h>
+#include <stdint.h>
```
Comment (sunghan-chang) [inline apps/examples/smtp_client/smtp_client.c]: same

File: apps/examples/smtp_client/smtp_client.c (hunk 14 above)
Comment (JeonginKim) [inline apps/examples/smtp_client/smtp_client.c]: I fixed it.

File: apps/examples/smtp_client/smtp_client.c (hunk 13 above)
Comment (JeonginKim) [inline apps/examples/smtp_client/smtp_client.c]: I fixed it.

File: apps/examples/smtp_client/smtp_client.c (hunk 12 above)
Comment (JeonginKim) [inline apps/examples/smtp_client/smtp_client.c]: I fixed it.

File: apps/examples/smtp_client/smtp_client.c (hunk 15 above)
Comment (JeonginKim) [inline apps/examples/smtp_client/smtp_client.c]: I fixed it.

**Pull request merged**